### PR TITLE
chore: migrate drawer

### PR DESCRIPTION
**Pull request by anucreative**
## SCREENSHOTS / SCREEN RECORDINGS

image

## COMPATIBILITY

- [ ] Tested on Safari (desktop)
- [x] Tested on Chrome (desktop)
- [ ] Tested on Firefox (desktop)
- [ ] Tested on mobile device sizes
- [ ] Tested on tablet device sizes
- [ ] Tested on IOS Safari (either device or simulator)

## QA

- [x] Thoroughly tested in local environment
- [ ] Added tests for all new features
- [ ] Added tests that considered edge cases

### CODE REVIEW

File: lib/src/components/Drawer/AssetDrawer/IconBlock.tsx [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+
+import styles from './asset-drawer.module.scss'
+import type { IconBlockProps } from './types'
+
+const cx = classNames(styles)
+
+export const IconBlock = ({ icon: Icon, size = 'md' }: IconBlockProps) => {
+  return (
+    <div className={cx('icon-block', `size-${size}`)}>
+      <Icon color="neutral-90" size={size} />
```
Comment (cnairi) [inline lib/src/components/Drawer/AssetDrawer/IconBlock.tsx]: Is the color working here? I think it shouldn't 🤔 (I don't see it handled in the Icon component)

File: lib/src/components/Drawer/AssetDrawer/asset-drawer.module.scss [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+    flex-shrink: 0;
+    border-radius: var(--radius-md);
+    background-color: var(--color-beige-20);
+    transition: background-color var(--timing-primary);
+
+    width: var(--icon-block-size);
+    height: var(--icon-block-size);
+
+    &.size-md {
+      --icon-block-size: 48px;
```
Comment (cnairi) [inline lib/src/components/Drawer/AssetDrawer/asset-drawer.module.scss]: It could be var(--height-48)?

File: lib/src/components/Drawer/AssetDrawer/asset-drawer.module.scss [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+    }
+
+    &.size-sm {
+      --icon-block-size: 32px;
```
Comment (cnairi) [inline lib/src/components/Drawer/AssetDrawer/asset-drawer.module.scss]: It could be var(--height-32)

File: lib/src/components/Drawer/components/Trigger.tsx [hunk 4]
```diff
@@ -0,0 +1,13 @@
+import { DialogDisclosure } from '@ariakit/react'
+
+import { forwardRefWithAs } from '@/utils/forwardRefWithAs'
+
+import type { TriggerProps } from '../types'
+
+export const Trigger = forwardRefWithAs<TriggerProps, 'button'>(({ as, store, ...rest }, ref) => {
+  const Element = as || 'button'
+
+  return (
```
Comment (cnairi) [inline lib/src/components/Drawer/components/Trigger.tsx]:
```suggestion
export const Trigger = forwardRefWithAs<TriggerProps, 'button'>(
  ({ as: Element = 'button', store, ...rest }, ref) => {
    return (
```

(it's a proposal for the as syntax)

File: lib/src/components/Drawer/index.tsx [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+import { Content } from './components/Content'
+import { Footer } from './components/Footer'
+import { Title } from './components/Title'
+import { Trigger } from './components/Trigger'
+import styles from './drawer.module.scss'
+import type { DrawerProps } from './types'
+
+const cx = classNames(styles)
+
+const DrawerComponent = forwardRefWithAs<DrawerProps, 'div'>(
```
Comment (cnairi) [inline lib/src/components/Drawer/index.tsx]: I don't think you need the as here? But I might be wrong

File: lib/src/components/Drawer/index.tsx [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+      withCloseButton = true,
+      ...rest
+    },
+    ref
+  ) => {
+    return (
+      <Dialog
+        backdrop={
+          withBackdrop ? (
+            <Backdrop hideOnInteractOutside={hideOnInteractOutside as boolean} />
```
Comment (cnairi) [inline lib/src/components/Drawer/index.tsx:39]: You can remove your cast here and modify your type: `export type BackdropOptions = { hideOnInteractOutside?: DialogOptions['hideOnInteractOutside'] }`

File: lib/src/components/Drawer/types.ts [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+  withCloseButton?: boolean
+}
+
+export type DrawerProps = ComponentPropsWithRef<'div'> & DrawerOptions
+
+export type FooterOptions = { className?: string }
+
+export type FooterProps = ComponentPropsWithRef<'div'> & FooterOptions
+
+export type TitleOptions = { className?: string }
```
Comment (cnairi) [inline lib/src/components/Drawer/types.ts]: I think you can remove all the additional className here, it's already in the div props

File: lib/src/components/Modal/modal.module.scss [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+    &.hideOnInteractOutside {
+      cursor: pointer;
+    }
+
+    &[data-enter] {
+      opacity: 1;
+    }
+  }
+
+  .root {
```
Comment (cnairi) [inline lib/src/components/Modal/modal.module.scss:4]: I would put the root at the top, before the backdrop, but this is optional

File: lib/src/components/Drawer/drawer.module.scss [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+          }
+          &-md {
+            width: 34.375rem; /* 550px */
+          }
+          &-lg {
+            width: 42.5rem; /* 680px */
+          }
+        }
+      }
+    }
```
Comment (cnairi) [inline lib/src/components/Drawer/drawer.module.scss:78]: From what I read, the size is not different according to the placement,
Could you just
```
    &.placement {
      &-bottom {
        bottom: 0;
        left: 0;
        right: 0;
        transform: translateY(100%);
      }
      &-left {
        bottom: 0;
        left: 0;
        top: 0;
        transform: translateX(-100%);
      }
      &-right {
        bottom: 0;
        right: 0;
        top: 0;
        transform: translateX(100%);
      }
      &-top {
        left: 0;
        right: 0;
        top: 0;
        transform: translateY(-100%);
      }
    }

     &.size {
      &-sm {
        width: 25rem; /* 400px */
      }
      &-md {
        width: 34.375rem; /* 550px */
      }
      &-lg {
        width: 42.5rem; /* 680px */
      }
    }
``` ?

File: lib/src/components/Drawer/AssetDrawer/Header.tsx [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+import { Text } from '@/components/Text'
+import { classNames } from '@/utils'
+
+import styles from './asset-drawer.module.scss'
+import { IconBlock } from './IconBlock'
+import type { HeaderProps } from './types'
+
+const cx = classNames(styles)
+
+export const Header = ({
```
Comment (Leiksa) [inline lib/src/components/Drawer/AssetDrawer/Header.tsx:12]: We should use `forwardRefWithAs` here so we can make it easily overridable

File: lib/src/components/Drawer/AssetDrawer/index.tsx [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+      children,
+      className,
+      getPersistentElements: parentGetPersistentElements,
+      maxWidth,
+      store,
+      ...rest
+    },
+    ref
+  ) => {
+    const getPersistentElements = () =>
```
Comment (Leiksa) [inline lib/src/components/Drawer/AssetDrawer/index.tsx]: It would be great if the two callback are memoized as we don't know how Ariakit handle their memoization and we don't want to create extra re-render/listener

File: lib/src/components/Drawer/AssetDrawer/types.ts [hunk 12]
```diff
@@ -0,0 +1,38 @@
...
+
+import type { IconProps } from '@/components/Icon'
+
+export interface AssetDrawerOptions extends DialogOptions<'div'> {
+  maxWidth?: number | string
+}
+
+export type AssetDrawerProps = AssetDrawerOptions & ComponentPropsWithRef<'div'>
+
+export type HeaderProps = ComponentProps<'div'> & {
```
Comment (Leiksa) [inline lib/src/components/Drawer/AssetDrawer/types.ts:12]: Not needed anymore with the `forwardRefWithAs`
```suggestion
export type HeaderProps = {
```

File: lib/src/components/Drawer/components/Close.tsx [hunk 13]
```diff
@@ -0,0 +1,15 @@
...
+
+import { CloseButton } from '@/components/CloseButton'
+
+import type { CloseProps } from '../types'
+
+export const Close = ({ className }: CloseProps) => {
+  return (
+    <DialogDismiss
+      render={props => (
+        <CloseButton className={`absolute right-lg top-lg z-10 ${className}`} {...props} />
```
Comment (Leiksa) [inline lib/src/components/Drawer/components/Close.tsx]: Could we use css styling here and not tailwind 🙈

Comment (Leiksa) [reply lib/src/components/Drawer/components/Close.tsx]: (Also if not, you should use the `classNames` utility otherwise you will have extra space usage when there are no `className` passed as props)

File: lib/src/components/Drawer/components/Content.tsx [hunk 14]
```diff
@@ -0,0 +1,12 @@
+import { forwardRef } from 'react'
+
+import { classNames } from '@/utils'
+
+import styles from '../drawer.module.scss'
+import type { ContentProps } from '../types'
+
+const cx = classNames(styles)
+
+export const Content = forwardRef<HTMLDivElement, ContentProps>(({ className, ...rest }, ref) => {
```
Comment (Leiksa) [inline lib/src/components/Drawer/components/Content.tsx:10]:
```suggestion
export const Content = forwardRefWithAs<ContentProps, 'div'>(({ className, ...rest }, ref) => {
```

File: lib/src/components/Drawer/components/Footer.tsx [hunk 15]
```diff
@@ -0,0 +1,12 @@
+import { forwardRef } from 'react'
+
+import { classNames } from '@/utils'
+
+import styles from '../drawer.module.scss'
+import type { FooterProps } from '../types'
+
+const cx = classNames(styles)
+
+export const Footer = forwardRef<HTMLDivElement, FooterProps>(({ className, ...rest }, ref) => {
```
Comment (Leiksa) [inline lib/src/components/Drawer/components/Footer.tsx:10]:
```suggestion
export const Footer = forwardRefWithAs<FooterProps, 'div'>(({ className, ...rest }, ref) => {
```

File: lib/src/components/Drawer/components/Title.tsx [hunk 16]
```diff
@@ -0,0 +1,17 @@
+import { Text } from '@/components/Text'
+import { classNames } from '@/utils'
+
+import styles from '../drawer.module.scss'
+import type { TitleProps } from '../types'
+
+const cx = classNames(styles)
+
+export const Title = ({ children, className, ...rest }: TitleProps) => {
```
Comment (Leiksa) [inline lib/src/components/Drawer/components/Title.tsx:9]: `forwardRefWithAs` 🙈

File: lib/src/components/Drawer/components/Trigger.tsx [hunk 17]
```diff
@@ -0,0 +1,13 @@
...
+    <DialogDisclosure ref={ref} render={props => <Element {...props} />} store={store} {...rest} />
```
Comment (Leiksa) [inline lib/src/components/Drawer/components/Trigger.tsx]: For ariakit component we don't want to run the extra render function so maybe we could just not default the as and just do `as: As` and the following (Ariakit will handle the props forwarding):

```suggestion
    <DialogDisclosure ref={ref} render={As ? <As /> : undefined} store={store} {...rest} />
```

File: lib/src/components/Drawer/index.tsx (hunk 5 above)
Comment (Leiksa) [reply lib/src/components/Drawer/index.tsx]: We need it but it is not used currently in the `render` so we must pass it

File: lib/src/components/Drawer/index.tsx [hunk 18]
```diff
@@ -0,0 +1,71 @@
...
+          ) : (
+            false
+          )
+        }
+        hideOnInteractOutside={hideOnInteractOutside}
+        modal={withBackdrop}
+        ref={ref}
```
Comment (Leiksa) [inline lib/src/components/Drawer/index.tsx:46]: We default the as to `div`:
```tsx
const Element: as || 'div'

``` 

```suggestion
        className={cx('root', `placement-${placement}`, `size-${size}`, className)}
        ref={ref}
        render={<Element />}
```

**Comment by anucreative** [comment]
> I spotted some differences with our current component:

This is because I updated the docs to include <Drawer.Content /> so that it looked "nicer".

File: lib/src/components/Drawer/index.tsx (hunk 6 above)
Comment (anucreative) [reply lib/src/components/Drawer/index.tsx:39]: `DialogOptions['hideOnInteractOutside']` is `boolean` or `function`

File: lib/src/components/Drawer/components/Close.tsx (hunk 13 above)
Comment (anucreative) [reply lib/src/components/Drawer/components/Close.tsx]: > Could we use css styling here and not tailwind 🙈

`CloseButton` is a WUI component so we want to use Tailwind to override styles (from what I understand)

File: lib/src/components/Drawer/components/Footer.tsx (hunk 15 above)
Comment (anucreative) [reply lib/src/components/Drawer/components/Footer.tsx:10]: Do we want to use `forwardRefWithAs` everywhere ?

File: lib/src/components/Drawer/AssetDrawer/IconBlock.tsx (hunk 1 above)
Comment (theo-mesnil) [reply lib/src/components/Drawer/AssetDrawer/IconBlock.tsx]: Need a className

File: lib/src/components/Drawer/components/Close.tsx (hunk 13 above)
Comment (Leiksa) [reply lib/src/components/Drawer/components/Close.tsx]: This is just my opinion, but I would prefer to use Tailwind overrides only outside of components (for example, in examples or in our project setup) and never within our actual implementation. 🤔

File: lib/src/components/Drawer/components/Footer.tsx (hunk 15 above)
Comment (Leiksa) [reply lib/src/components/Drawer/components/Footer.tsx:10]: Again in my opinion 🙈  -> it’s a missed opportunity not to leverage this flexibility, so we should make use of it wherever possible

File: lib/src/components/Drawer/components/Close.tsx (hunk 13 above)
Comment (cnairi) [reply lib/src/components/Drawer/components/Close.tsx]: cf => https://wttj.slack.com/archives/C09CT2BSF0B/p1758034277867449 In this thread, we said it was acceptable to use tailwind to compound our components when we have no other choice, I think here we can create a class in css modules to override as it's 100% static

File: lib/src/components/Drawer/components/Footer.tsx (hunk 15 above)
Comment (cnairi) [reply lib/src/components/Drawer/components/Footer.tsx:10]: I think @theo-mesnil has another opinion on that topic, it's to allow the as only when necessary

Comment (theo-mesnil) [reply lib/src/components/Drawer/components/Footer.tsx:10]: On some components we don't want to have an as option, like label for example :) or other component with fixed element

File: lib/src/components/Drawer/components/Close.tsx (hunk 13 above)
Comment (Leiksa) [reply lib/src/components/Drawer/components/Close.tsx]: Yep, from what I understand we should use tailwind only when we have no other choice and remain at most to scss. Is that right?

File: lib/src/components/Drawer/components/Footer.tsx (hunk 15 above)
Comment (anucreative) [reply lib/src/components/Drawer/components/Footer.tsx:10]: For me, I don't see the need for the Footer / Content to use `as`. Alors, I'm with @theo-mesnil – only use `as` when needed.

File: lib/src/components/Drawer/drawer.module.scss (hunk 9 above)
Comment (anucreative) [reply lib/src/components/Drawer/drawer.module.scss:78]: The `height` / `width` changes depending on whether the `placement` is `top` / `bottom` (`width` changes) or `left` / `right` (`height` changes)

File: lib/src/components/Drawer/AssetDrawer/index.tsx [hunk 19]
```diff
@@ -43,8 +51,9 @@ export const AssetDrawerComponent = forwardRefWithAs<AssetDrawerProps, 'div'>(
         hideOnInteractOutside={hideOnInteractOutsideFn}
         placement="bottom"
         ref={ref}
-        size={null}
         store={store}
+        // Needed to override the default height set in drawer.module.scss
+        style={{ '--drawer-height': 'calc(100% - 3rem)' } as React.CSSProperties}
```
Comment (Leiksa) [inline lib/src/components/Drawer/AssetDrawer/index.tsx:56]: You could override it directly in the `asset-drawer` class
```
.asset-drawer {
...css
--drawer-height: calc(100% - 3rem)
}
```

Comment (anucreative) [reply lib/src/components/Drawer/AssetDrawer/index.tsx:56]: I did that at first but it didn't work because sometimes `.drawer.size-md` styles have a higher specificity than `.asset-drawer`. 

I agree that this looks like is a bit of a hack – but it works and isn't _too_ ugly. 🤓
